tchannel/tornado/peer.py

```diff
@@ -540,121 +551,100 @@ def send(
                 name=endpoint,
                 trace_id=trace_id,
                 parent_span_id=parent_span_id,
-                endpoint=Endpoint(self.peer.host,
-                                  self.peer.port,
-                                  self.service),
+                endpoint=Endpoint(peer.host, peer.port, self.service),
                 traceflags=traceflag,
             )
         )
 
         # only retry on non-stream request
         if request.is_streaming_request or self._hostport:
-            attempt_times = 1
+            retry_limit = 1
 
         if request.is_streaming_request:
             request.ttl = 0
 
-        # event: send_request
-        self.peer.tchannel.event_emitter.fire(
-            EventType.before_send_request,
-            request,
-        )
-
         try:
             response = yield self.send_with_retry(
-                request, peer, attempt_times
-            )
-        except ProtocolError as protocol_error:
-            # event: after_receive_protocol_error
-            self.peer.tchannel.event_emitter.fire(
-                EventType.after_receive_error,
-                request,
-                protocol_error,
+                request, peer, retry_limit, connection
             )
-            raise
         except Exception as e:
-            # event: on_operational_error
-            self.peer.tchannel.event_emitter.fire(
-                EventType.on_exception,
-                request,
-                e,
+            # event: on_exception
+            self.tchannel.event_emitter.fire(
+                EventType.on_exception, request, e,
             )
             raise
 
         log.debug("Got response %s", response)
-        # event: after_receive_response
-        self.peer.tchannel.event_emitter.fire(
-            EventType.after_receive_response,
-            request,
-            response,
-        )
 
         raise gen.Return(response)
 
     @gen.coroutine
     def _send(self, connection, req):
+        # event: send_request
+        self.tchannel.event_emitter.fire(EventType.before_send_request, req)
         response_future = connection.send_request(req)
+
         with timeout(response_future, req.ttl):
-            response = yield response_future
+            try:
+                response = yield response_future
+            except ProtocolError as protocol_error:
```

Comment:
Currently ProtocolError represent Error frames we received from network. We use code to know the specific error message. We will make different Errors though